### PR TITLE
Rename endof->lastindex and introduce firstindex. Fixes #23354

### LINKED ISSUE #23354: beginof(a) analogue to endof(a)?
Looping over `1:length(a)` is wrong for `AbstractVector` because it fails for non 1-based arrays.  `eachindex(a)` is correct, but occasionally you need to provide explicit indices, e.g. to loop only over odd indices as in #23267.  `start(a):endof(a)` doesn't work because for views `start(a)` is a tuple.   You can do `first(eachindex(a))`, but this is kind of klunky and non-obvious.

Can we export a function ~~`startof`~~ `beginof(a) = first(eachindex(a))` analogous to `endof(a)`?

### PR DESCRIPTION
**Pull request by timholy**
While working on https://github.com/JuliaLang/Juleps/issues/47 I noticed that we really should have `beginindex` (see #23354). This *almost* resolves that issue: I think I have all the tedious stuff done, but my lisp is laughable and character-twiddling was insufficient to fix the parsing of `a[begin]`. Would love some help. @stevengj? (Anyone who wishes, you are more than welcome to push directly to this branch, or do whatever you find most convenient.)

I defined specific `beginindex` methods for every type that defines what is now called `endindex` (formerly `endof`). Since `beginindex` is a new function, I also added a "deprecation" that returns 1 for any object, as a way of encouraging package authors to define it as needed. Hope this seems like a reasonable strategy.

### CODE REVIEW

File: src/julia-syntax.scm [hunk 1]
```diff
@@ -86,35 +86,46 @@
...
 ;; `last` = is this last index?
-;; returns a call to endof(a) or size(a,n)
+;; returns a call to, e.g., endindex(a) or last(axes(a,n))
 (define (end-val a n tuples last)
   (if (null? tuples)
       (if (and last (= n 1))
-          `(call (top endof) ,a)
-          `(call (top size) ,a ,n))
+          `(call (top endindex) ,a)
+          `(call (top last) (call (top axes) ,a ,n)))
```
Comment (mbauman) [inline src/julia-syntax.scm]: I like this a lot, but could we incorporate this behavior into `endindex`, too?  That way the parser only lowers `end` to `endindex(A)` or `endindex(A, n)` for the one- and n-arg cases, respectively.

**Comment by stevengj** [comment]
Can we just keep `endof` and call it `startof`?

**Comment by ararslan** [comment]
or `firstof`?

**Comment by timholy** [comment]
Keeping it `endof` is of course an option, just not what appeared to be decided in #23354. I'm happy to go back to `endof`, though.

If we implement `a[begin]` I don't think it makes sense to call it `firstof`.

If someone gets to the parsing portion of this soon, great. If not, perhaps it would make sense to merge with `beginof`/`beginindex` and then implement `a[begin]` in 1.x.

**Comment by timholy** [comment]
Note that the only part of this that would have to be done for 1.0 is the rename, which is part of why I tackled this.

**Comment by GregPlowman** [comment]
Here's a user's perspective: 

> Can we just keep endof and call it beginof?

I think `beginindex` and `endindex` more clearly conveys that an _index_ rather than an _element_ is returned.
I used to get confused between `endof(collection)` and `last(collection)`


> or firstof?

I don't think it ties in as nicely with `a[begin]`
https://github.com/JuliaLang/julia/issues/23354#issuecomment-323546576

**Comment by Sacha0** [comment]
Echoing `@GregPlowman`, the `index` names seem more descriptive.

Though `firstindex`/`lastindex` are less consistent with `begin`/`end` than `beginindex`/`endindex`, `firstindex`/`lastindex` are more consistent with `first`/`last` than `beginindex`/`endindex`. Additionally, `firstindex`/`lastindex` are IMHO more descriptive/intuitive besides. (`beginindex`/`endindex` suggest performing "begin" and "end" operations on an index, whatever that would mean, rather than getting the first and last indices of a collection).

Intuitiveness in direct/explicit use of `firstindex`/`lastindex` (and particularly alongside `first`/`last`) seems like a more meaningful consideration than consistency between `begin`/`end` and the forms those lower to. Best! (Apologies for wading into the bikeshed!)

**Comment by stevengj** [comment]
One difficulty with parsing a[begin] is that some look ahead is required, since a[begin...end] is valid.  

Maybe if `begin` is encountered in a ref expression, it should only be treated as an identifier if it is followed by an operator, comma, or ]?

**Comment by timholy** [comment]
Maybe also preceded? e.g. `a[2*begin:end]`.

**Comment by mbauman** [comment]
When I had thought about doing this in the past, I was thinking we could just deprecate `begin … end` blocks within indexing altogether.  We could also just break it entirely if there's enough weight to enabling this syntax relative to the number of uses of `x[begin … end]`.

I really don't think we should try to support both `begin`/`end` as indices and `begin … end` blocks at the same time.

**Comment by stevengj** [comment]
@mbauman, yes, you are probably right.  For example, `a[begin +1; end]` is currently valid syntax (equivalent to `a[1]`), but it means that we can't tell purely from the subsequent character whether `begin` is the start of a block or should be used as a variable.

**Comment by yurivish** [comment]
What would happen if the interior of an indexing expression contains a macro call that expands to a begin/end block?

Relatedly, would cases like `a[f(begin 1 + 1 end)]` still be allowed?

**Comment by mauro3** [comment]
Could we still rename `begin ... end` blocks?  `block ... end` would be reasonable.

**Comment by KristofferC** [comment]
`block ... end` is quite appealing. `for` begins a for loop, `if` begins an if statement, `block` begins a block. `begin` is pretty non-saying, begin what?

**Comment by stevengj** [comment]
@yurivish, we are discussing the parser, which affects only surface syntax.  Macro expansion happens later, and wouldn't be affected.

**Comment by StefanKarpinski** [comment]
I rather like `endof` and `beginof` but I certainly hear (and agree) with what @GregPlowman is saying about `endindex` and `beginindex` being much more clearly evocative of what they mean.

`block ... end` is not bad at all, but I also think that disallowing `begin ... end` inside of indexing is a pretty sane thing to do. This is already a bit of a problem since you cannot use the index meaning of `end` inside of a `begin ... end` block inside of an indexing expression (reasonably enough).

**Comment by Keno** [comment]
I don't have much to add here at the moment, except to add that in #25261 I ran into a bunch of problems with code assuming that the iteration and indexing states where identical (e.g. using `start`/`endof` to bound indices and `next` instead of `nextindex`), so I'm happy to see this.

**Comment by stevengj** [comment]
(`begin` probably comes from Scheme's `(begin ...)` construct.)   It seems quite annoying to rename a basic keyword of the language, and to introduce a new reserved word `block`, at this stage of the game.

**Comment by timholy** [comment]
I'm fine with any reasonable solution to the parsing issues.

**Comment by StefanKarpinski** [comment]
I think the best approach is probably to disallow `begin ... end` blocks inside of indexing in 0.7 and then introduce `a[begin]` as an analogue to `a[end]` in 1.0. The names `beginindex` and `endindex` seem fine to me.

**Comment by JeffBezanson** [comment]
I do prefer the names `beginindex` and `endindex` to the current names. If this PR is changed just to do those renamings, I can add a deprecation for `begin ... end` inside indexing, and then we can add `a[begin]` safely in 1.0.

**Comment by Keno** [comment]
Needs a rebase. If the syntax part of this is controversial, maybe we can at least get the `beginindex` parts? I need to separate indexing state from iteration before we can merge #25261, so this would be a good thing to base that on.

**Comment by Keno** [comment]
Actually I'll just go ahead, rebase this, and take out the syntax parts (hope you don't mind).

**Comment by Keno** [comment]
Rebased. I backed up the `begin` syntax as a separate commit, so whoever wants to take that up after 1.0 can just revert that change.

File: doc/src/manual/interfaces.md [hunk 2]
```diff
@@ -161,7 +161,8 @@ julia> collect(Iterators.reverse(Squares(10)))' # transposed to save space
 |:-------------------- |:-------------------------------- |
 | `getindex(X, i)`     | `X[i]`, indexed element access   |
 | `setindex!(X, v, i)` | `X[i] = v`, indexed assignment   |
-| `endof(X)`           | The last index, used in `X[end]` |
+| `beginindex(X)`         | The first index, used in `X[begin]` |
```
Comment (mbauman) [inline doc/src/manual/interfaces.md]: The syntax here is a holdover.

**Comment by gabrielgellner** [comment]
I find `beginindex` is really hard to scan due to the internal `inin` this feels like an future api wart to me.

**Comment by StefanKarpinski** [comment]
We could just call it `begindex` :trollface:

**Comment by fredrikekre** [comment]
... and `endex`. `begindex` and `endex` seems like a nice couple 😄

**Comment by KristofferC** [comment]
> `begindex` and `endex` seems like a nice couple

Too bad they are ex:es.

**Comment by Keno** [comment]
Rebased again. I'm planning to merge this one CI passes.

**Comment by StefanKarpinski** [comment]
Since `beginindex` is indeed a horrific name and when we discussed this previously, everyone agreed that `firstindex` and `lastindex` were the clearest names semantically, let's go with those instead. It's not so terrible that `a[end]` calls `lastindex` and `a[begin]` calls `firstindex`. `lastindex` is a far clearer name than `endof` for example, which was fairly opaque.

**Comment by JeffBezanson** [comment]
I agree. It's not really confusing for `a[end]` to call `lastindex`. After all, it's also not obvious that `a[end]` would correspond to `endof` or `endindex`, or that `.` would call `broadcast`, etc.

**Comment by Keno** [comment]
Rebased and renamed.

**Comment by ararslan** [comment]
You might want to amend the commit message for the first commit here since what it describes is no longer quite what it does, since the names are different.

File: NEWS.md [hunk 3]
```diff
@@ -991,6 +991,8 @@ Deprecated or removed
 
   * `Base.@gc_preserve` has been deprecated in favor of `GC.@preserve` ([#25616]).
 
+  * `endof(a)` has been renamed `endindex(a)` ([#23554]).
```
Comment (StefanKarpinski) [inline NEWS.md]: `lastindex`

Comment (stevengj) [reply NEWS.md]: "renamed to"

**Comment by Keno** [comment]
Some idiot merged something that conflicts with this. I'll rebase again, sigh. Planning to merge next time we're green on CI.

**Comment by Keno** [comment]
People expressed concern about the phrasing, so for the record, it was me in #25733, I'm the idiot ;).

**Comment by timholy** [comment]
Thanks for picking this up and pushing it over the line, @Keno; it's been a busy few weeks.

**Pull request merged**